src/Workspaces/Core/Portable/Shared/Extensions/INamespaceOrTypeSymbolExtensions.cs

```diff
@@ -82,5 +83,33 @@ private static void GetNameParts(INamespaceOrTypeSymbol namespaceOrTypeSymbol, L
             GetNameParts(namespaceOrTypeSymbol.ContainingNamespace, result);
             result.Add(namespaceOrTypeSymbol.Name);
         }
+
+        /// <summary>
+        /// Lazily returns all nested types contained (recursively) within this namespace or type.
+        /// In case of a type, it is included itself as the first result.
+        /// </summary>
+        public static IEnumerable<INamedTypeSymbol> GetAllTypes(
```

Comment:
The body of this method is exactly the same, just I moved it from `INamespaceSymbolExtensions` and changed the parameter to `INamespaceOrTypeSymbol`.

Reply:
Maybe this can be optimized a bit with a pre-filter, rather than post-filtering.

We could check `IsAccessibleWithin` here, rather than after, with the following reasoning:
```
class A
{
    class C
    {
        class MyAttribute : Attribute {}
    }
}

class B
{
    [$
    public void MyMethod() {}
}
```

Where $ is the caret where we trigger completion. The current method would unroll C and members of C and iterate them, which can cause additional stack resizes.

A minor change could be done as:
i.e.
```
public static IEnumerable<INamedTypeSymbol> GetAllTypes<T>(
    this INamespaceOrTypeSymbol namespaceOrTypeSymbol,
    CancellationToken cancellationToken,
    Func<INamedTypeSymbol, T, bool> filter = null,
    T state = default)
{
    var stack = new Stack<INamespaceOrTypeSymbol>();
    stack.Push(namespaceOrTypeSymbol);

    while (stack.Count > 0)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var current = stack.Pop();
        if (current is INamespaceSymbol currentNs)
        {
            stack.Push(currentNs.GetMembers());
        }
        else
        {
            var namedType = (INamedTypeSymbol)current;
            if (filter != null && filter (namedType, state))
            {
                stack.Push(namedType.GetTypeMembers());
                yield return namedType;
            }
        }
    }
}

...
// Usage here. Note that we only do the `withinType ?? withinAssembly` check once now.
foreach (var type in namespaceOrType.GetAllTypes(cancellationToken, (t, within) => t.IsAccessibleWithin (within), withinType ?? withinAssembly)) {
    if (type.IsAttribute())
    {
        return true;
    }
}
```

Reply:
Or, instead of optional parameters, another overload could be added. It might be useful in other scenarios, not just this one.

Reply:
@Therzok Thanks, that look good like a possible way to do things but I wouldn't do this without measuring the difference first either. The concern I have here is that `IsAccessibleWithin` could potentially be a lot more expensive than `IsAttribute`, which would make this slower. When I dig a little into the implementation, it definitely looks more complex. `IsAttribute`, on the other hand is a simple base class check.

Reply:
I was going through the source code and noticed this, which is similar:
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/FullyQualify/AbstractFullyQualifyCodeFixProvider.cs,298

There's also this, but it's a bit different, as it would need:
a) a filter which decides whether to yield return
b) a filter which decides whether to push its members

http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Debugging/AbstractBreakpointResolver.cs,141
http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Debugging/AbstractBreakpointResolver.cs,214
^ this one is pretty indirect, walk up the call tree from here.

Reply:
So, a final catch-all proposal would be:
```
public static IEnumerable<INamedTypeSymbol> GetAllTypes<T>(
    this INamespaceOrTypeSymbol namespaceOrTypeSymbol,
    CancellationToken cancellationToken,
    Func<INamedTypeSymbol, T, bool> resultFilter = null,
    T resultState = default,
    Func<INamedTypeSymbol, U, bool> childrenFilter = null,
    U childrenState = default)
{
    var stack = new Stack<INamespaceOrTypeSymbol>();
    stack.Push(namespaceOrTypeSymbol);

    while (stack.Count > 0)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var current = stack.Pop();
        if (current is INamespaceSymbol currentNs)
        {
            stack.Push(currentNs.GetMembers());
        }
        else
        {
            var namedType = (INamedTypeSymbol)current;
            if (childrenFilter == null || childrenFilter (namedType, childrenState))
            {
                stack.Push(namedType.GetTypeMembers());
            }
            if (resultFilter == null || resultFilter (namedType, resultState))
            {
                yield return namedType;
            }
        }
    }
}

Reply:
I think we'd prefer the solution with another overload since the same functionality is already used in another place. But as I said, we don't know if this would actually be better given how complex IsAccessibleWithin is compared to IsAttribute.

Reply:
I think this can be modeled in a way that we can:
a) both gain performance of not pushing in types which are not accessible
b) use the faster IsAttribute() check for result filtering.

Reply:
@Therzok But childrenFilter (IsAccessibleWithin) is still called before IsAttribute, which *might* be a problem, depending on how efficient it is.

Reply:
Okay, I misunderstood what you meant by running `IsAccessibleWithin` on every node. Yes, doing post-filtering is more beneficial now.